rpcs3/Loader/TAR.cpp

```diff
@@ -2,38 +2,48 @@
 
 #include "TAR.h"
 
+#include "util/asm.hpp"
+
 #include <cmath>
 #include <cstdlib>
+#include <charconv>
 
 LOG_CHANNEL(tar_log, "TAR");
 
-tar_object::tar_object(const fs::file& file, usz offset)
+tar_object::tar_object(const fs::file& file)
 	: m_file(file)
-	, initial_offset(static_cast<int>(offset))
 {
-	m_file.seek(initial_offset);
-	largest_offset = initial_offset;
 }
 
-TARHeader tar_object::read_header(u64 offset)
+TARHeader tar_object::read_header(u64 offset) const
 {
-	m_file.seek(offset);
-	TARHeader header;
-	m_file.read(header);
+	TARHeader header{};
+
+	if (m_file.seek(offset) != offset)
+	{
+		return header;
+	}
+
+	if (!m_file.read(header))
+	{
+		std::memset(&header, 0, sizeof(header));
+	}
+
 	return header;
 }
 
-int octalToDecimal(int octalNumber)
+u64 octal_text_to_u64(std::string_view sv)
 {
-	int decimalNumber = 0, i = 0, rem;
-	while (octalNumber != 0)
+	u64 i = -1;
+	const auto [ptr, err] = std::from_chars(sv.data(), sv.data() + sv.size(), i, 8);
```

Comment:
from_chars allows much, much better error checking than atoi which is perfect for this. If theres any error `i` will remain UINT64_MAX which fails the later checks.